✨ Add Spring Boot MicroShift product

mcarlett opened the pull request:
Adds 

- new product `springboot-microshift` 
- new deployment strategy `FromImageStrategy` that creates Docker image from generated app and push it to the remote registry, than create a Kubernetes resources (deployment, config map and service)
- system-x services implemented : `AMQ Broker` , `IBM MQ`, `Infinispan`, `MySql`

mcarlett commented:
> @mcarlett can you please elaborate a bit how the microshift stuff should work? (I have no experience with that)
> 
> at the first glance microshift should be a subset of openshift, so what are we missing from openshift features that require a whole new `product` class?
> 
> what's the difference in the system-x service classes? they seem to be 99% identical to the openshift system-x services
> 
> since we control the archetype, wouldn't it be easier to delegate all the docker stuff to some maven plugin, instead of doing it in the code?

@avano I implemented a new deploy strategy to keep separated the way to deploy the product. Yes this is not the only way to build docker image and push it in a registry (MicroShift doesn't have Build resource), I can do something similar using jkube and I planned to implement another deployment strategy based on that. I added new product to handle different scenarios and the current implementations of system-x-services for Openshift are not compatible (they use DeploymentConfig) and I thought to keep separated, but if you think the solution to use Openshift implementations can be adopted, I can keep just a property and handle it into the implementations (ex: test.use.microshift will remain in the springboot-openshift product but I will add some conditions based on this property)

mcarlett commented:
> since we will need to migrate to deployments anyway, it's probably not worth duplicating the resources

yes indeed, I removed the product and I replacing deploymentconfig with deployment in the system-x-services I want to use for testing MicroShift

> what about the docker stuff? can't it be moved to some maven plugin in the archetype?

as I said it is possible to use the openshift maven plugin to push the generated image, I will implement this additional deployment strategy, do you think the current implemented cannot be used at all?

> also did you consider running own docker registry on the microshift, to avoid pushing all the images to quay? would that be possible?

out of the box MicroShift doesn't have it's own registry and I think installing a new one is not a good practice since MicroShift is built to run on edge devices with limited resources, so I would keep as light as possible, that's why I thought to use external image registry

thanks @avano

avano commented:
> > what about the docker stuff? can't it be moved to some maven plugin in the archetype?
> 
> as I said it is possible to use the openshift maven plugin to push the generated image, I will implement this additional deployment strategy, do you think the current implemented cannot be used at all?

well, of course it can be used, but I think it is not necessary to do all the stuff manually when some maven plugin can do the same with less hassle

> 
> > also did you consider running own docker registry on the microshift, to avoid pushing all the images to quay? would that be possible?
> 
> out of the box MicroShift doesn't have it's own registry and I think installing a new one is not a good practice since MicroShift is built to run on edge devices with limited resources, so I would keep as light as possible, that's why I thought to use external image registry

I think the only prerequisite is that you need to have *a* registry and it doesn't matter where/what the registry is. keep in mind that running the whole TNB testsuite builds 250+ docker images, so using some external image registry and cleaning it afterwards is just more work to do, that you can avoid by having ephemeral registry on the microshift instance

mcarlett commented:
> well, of course it can be used, but I think it is not necessary to do all the stuff manually when some maven plugin can do the same with less hassle

ok since I already implemented the usage of the embedded openshift-maven-plugin with external repository, I can remove the deprecated deployment strategy

> I think the only prerequisite is that you need to have _a_ registry and it doesn't matter where/what the registry is

this is on of the goal of the PR, so that I added a dedicated configuration parameter `openshift-sb.result-image.repo`, it is possible to configure this parameter based on the CI 

> . keep in mind that running the whole TNB testsuite builds 250+ docker images, so using some external image registry and cleaning it afterwards is just more work to do, that you can avoid by having ephemeral registry on the microshift instance

I don't think having an image repository in the microshift device is a good practice because usually the microshift cluster is installed in small devices with limited resources but, as in the previous point, TNB framework doesn't care the management of the registry since it handles only the configuration of the registry. The CI process will handle the registry

mcarlett commented on common/src/main/java/software/tnb/common/config/OpenshiftConfiguration.java:
```diff
@@ -25,6 +27,37 @@ public class OpenshiftConfiguration extends Configuration {
+    public static final String USE_MICROSHIFT = "test.use.microshift";
+
+    /**
+     * the port configured as node port in case of microshift is installed as VM in OCP Virtualization, on OCP cluster
```
not more necessary, I remove it, thanks
